### PR TITLE
feat: ems infrastructure

### PR DESCRIPTION
**Pull request by rahulguptajss**
TODO

1. verify ems template (add more exports)
2. Bookend events

### CODE REVIEW

File: cmd/collectors/ems/templating.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		name = strings.TrimSpace(values[0])
+		display = strings.TrimSpace(values[1])
+	} else {
+		name = rawName
+		display = strings.ReplaceAll(rawName, ".", "_")
+		display = strings.ReplaceAll(display, "-", "_")
+	}
+
+	return name, display, "label"
+}
```
Comment (Hardikl) [inline cmd/collectors/ems/templating.go:99]: can reuse ParseMetric from util.go

File: cmd/collectors/ems/templating.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+//LoadEmsPlugins loads built-in plugins or dynamically loads custom plugins
+//and adds them to the collector
+func (e *Ems) LoadEmsPlugins(params *node.Node) ([]plugin.Plugin, error) {
```
Comment (Hardikl) [inline cmd/collectors/ems/templating.go:103]: almost same as collector class, can be re-used after modification in last few lines in collector.go

File: cmd/collectors/ems/ems.go [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+
+func (e *Ems) InitCache() error {
+
+	var (
+		events *node.Node
+		err    error
+	)
+
+	if x := e.Params.GetChildContentS("object"); x != "" {
+		e.Prop.Object = x
```
Comment (Hardikl) [inline cmd/collectors/ems/ems.go:135]: e.object

File: cmd/collectors/ems/ems.go [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+	} else {
+		e.Prop.Object = strings.ToLower(e.Object)
```
Comment (Hardikl) [inline cmd/collectors/ems/ems.go:137]: e.object

File: cmd/collectors/ems/ems.go [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+	if e.Plugins == nil {
+		e.Plugins = make(map[string][]plugin.Plugin)
+	}
+
+	for _, line := range events.GetChildren() {
+		prop := emsProp{}
+
+		prop.InstanceKeys = make([]string, 0)
+		prop.InstanceLabels = make(map[string]string)
+		prop.Counters = make(map[string]string)
```
Comment (Hardikl) [inline cmd/collectors/ems/ems.go:175]: what's the use-case of this?

**Comment by Hardikl** [comment]
codeflow would be working for existing Rest plugin as well as newly added(dummy) ems abc plugin.
```
hardikl@hardikl-mac-0 harvest % ./bin/poller --poller cluster-07 --loglevel 2 --promPort 12991 -c Ems --config harvest.yml 
6:05PM INF ./poller.go:184 > log level used: info Poller=cluster-07
6:05PM INF ./poller.go:185 > options config: harvest.yml Poller=cluster-07
6:05PM INF ./poller.go:222 > started in foreground [pid=17378] Poller=cluster-07
6:05PM INF github.com/netapp/harvest/v2/cmd/collectors/rest/rest.go:175 > Using timeout Poller=cluster-07 collector=Ems:Ems timeout=1m0s
6:05PM INF github.com/netapp/harvest/v2/cmd/poller/collector/helpers.go:123 > best-fit template [conf/ems/9.6.0/ems.yaml] for [9.11.1] Poller=cluster-07 collector=Ems:Ems
6:05PM WRN github.com/netapp/harvest/v2/cmd/collectors/ems/ems.go:131 > no ems plugin found  Poller=cluster-07 collector=Ems:Ems kind=abc
6:05PM INF github.com/netapp/harvest/v2/cmd/collectors/ems/ems.go:283 >  Poller=cluster-07 cluster time zone=America/New_York collector=Ems:Ems
6:05PM INF ./poller.go:349 > Autosupport disabled since poller not connected to ONTAP. Poller=cluster-07 poller=cluster-07
6:05PM INF ./poller.go:354 > poller start-up complete Poller=cluster-07
6:05PM INF ./poller.go:502 > updated status, up collectors: 1 (of 1), up exporters: 1 (of 1) Poller=cluster-07
6:05PM INF github.com/netapp/harvest/v2/cmd/collectors/ems/ems.go:403 > Collected Poller=cluster-07 apiTime=2.710510498s collector=Ems:Ems dataPoints=7 instances=1 parseTime=0s
```

File: cmd/collectors/ems/ems.go [hunk 6]
```diff
@@ -363,23 +417,31 @@ func (e *Ems) PollData() (map[string]*matrix.Matrix, error) {
...
-		if j > len(names) {
-			j = len(names)
+	// build hrefs based on urlsize
+	var hrefs []string
+	start := 0
+	for end := 0; end < len(e.eventNames); end += 1 {
+		h := e.getHref(e.eventNames[start:end], filter)
+		if len(h) > e.maxUrlSize {
+			if end == 0 {
+				return nil, fmt.Errorf("url size is too small to form queries: %d", e.maxUrlSize)
```
Comment (Hardikl) [inline cmd/collectors/ems/ems.go:427]: logs quite confusing, I believe url size is large even without adding ems name, right ?

**Review by cgrinds — COMMENTED** [review]
nice job Rahul

**Comment by cgrinds** [comment]
Few general comments
- `getClusterTimeZone` should probably be on the same schedule as reading the catalog
- In `PollData` instead of removing instances from the "last poll except the parent object". It might be faster to create a new map and copy only the parent?

**Review by cgrinds — APPROVED** [review]

**Pull request merged**